ui/src/router.tsx

```diff
@@ -4,90 +4,68 @@ import Edit from "edit";
+  // bar to consistently use "../.." to navigate to the logical parent. The
+  // current React router ".." behavior is inconvenient because it navigates
+  // from index routes to their direct parent, which is the same effective
+  // route, so its behavior is different depending on whether an element is at a
+  // leaf route or not.
   return (
     <ErrorBoundary FallbackComponent={NotFound}>
-      <Switch>
-        {Object.entries(pages).map(([page, component]) => (
-          <Route exact key={page} path={page} component={component} />
-        ))}
-        <Route path="/">
-          <NotFound error={new PathError("Invalid URL.")} />
+      <Routes>
+        <Route index element={<UnderlaySelect />} />
+        <Route path=":underlayName">
+          <Route index element={<Datasets />} />
+          <Route path="cohorts/:cohortId">
+            <Route index element={<Overview />} />
+            <Route path="edit/:groupId/:criteriaId">
+              <Route index element={<Edit />} />
+            </Route>
+          </Route>
+          <Route path="conceptSets/:conceptSetId">
+            <Route index element={<ConceptSetEdit />} />
+          </Route>
         </Route>
-      </Switch>
+        <Route
+          path="*"
+          element={<NotFound error={new PathError("Invalid URL.")} />}
+        />
+      </Routes>
     </ErrorBoundary>
   );
 }
 
-// useParentUrl returns the second longest matching page, assuming that the
-// longest is the current page since longer ones wouldn't match.
-export function useParentUrl(): string | undefined {
-  const history = useHistory();
-  const page = Object.keys(pages)
-    .filter((page) => matchPath(history.location.pathname, { path: page }))
-    .sort((a, b) => b.length - a.length)[1];
-  return replaceParams(page, false, useParams());
+export function underlayURL(underlayName: string) {
+  return underlayName;
 }
 
-export type UrlParams = {
-  underlayName?: string;
-  cohortId?: string;
-  groupId?: string;
-  criteriaId?: string;
-  conceptSetId?: string;
-};
+export function cohortURL(cohortId: string) {
+  return "cohorts/" + cohortId;
+}
 
-export function createUrl(params: UrlParams): string {
-  for (const page of Object.keys(pages)) {
-    const url = replaceParams(page, true, params);
-    if (url) {
-      return url;
-    }
-  }
-  throw new PathError(`No path matches "${JSON.stringify(params)}".`);
+export function conceptSetURL(conceptSetId: string) {
+  return "conceptSets/" + conceptSetId;
 }
 
-function replaceParams(
-  path: string,
-  validate: boolean,
-  params: Record<string, string>
-) {
-  let count = 0;
-  const url = path?.replaceAll(/:(\w+)/g, (_, match) => {
-    count = params[match] && count >= 0 ? count + 1 : -1;
-    return params[match];
-  });
-  return !validate || count == Object.keys(params).length ? url : undefined;
+export function criteriaURL(groupId: string, criteriaId: string) {
```

Comment:
Is `groupId` the criteria type (e.g. condition, device)? An example input here as a comment would be helpful.

Reply:
Group IDs are random strings that identify the included and excluded groups of criteria. IDs are used instead of indexes or names to more easily support reordering, renaming, etc. The overall data structure is in `cohort.ts`. I don't think an example of random IDs would be particularly enlightening here.